File: Olympiad/COCI/official/2008/contest4_solutions/poklon.cpp
#include <algorithm>
#include <cstdio>
#include <cstring>
#include <vector>

using namespace std;

#define MAX 1000000
#define MAXN 100000

struct interval { int lo, hi; };
bool operator < ( const interval &A, const interval &B ) {
   if( A.lo != B.lo ) return A.lo > B.lo;
   if( A.hi != B.hi ) return A.hi < B.hi;
   return 0;
}

int n;
vector<interval> intervals;
int how[MAXN];
int dp[MAXN];

struct fenwick {
   int a[MAX+1];

   fenwick() { memset( a, -1, sizeof a ); }

   int query( int X ) {
      int ret = -1;
      for( int x = X; x > 0; x -= x&-x )
         if( a[x]!=-1 && (ret==-1 || dp[a[x]]>dp[ret]) ) ret = a[x];
      return ret;
   }

   void update( int X, int i ) {
      for( int x = X; x <= MAX; x += x&-x )
         if( a[x] == -1 || dp[i] > dp[a[x]] ) a[x] = i;
   }
} L;

int main( void ) {
   scanf( "%d", &n );

   for( int i = 0; i < n; ++i ) {
      interval I;
      scanf( "%d%d", &I.lo, &I.hi );
      intervals.push_back( I );
   }
   sort( intervals.begin(), intervals.end() );

   int ret = 0;

   for( int i = 0; i < n; ++i ) {
      dp[i] = 1;
      how[i] = -1;

      int x = L.query( intervals[i].hi );

      if( x != -1 ) {
         dp[i] = dp[x] + 1;
         how[i] = x;
      }

      L.update( intervals[i].hi, i );

      if( dp[i] > dp[ret] ) ret = i;
   }

   printf( "%d\n", dp[ret] );
   while( ret != -1 ) {
      printf( "%d %d\n", intervals[ret].lo, intervals[ret].hi );
      ret = how[ret];
   }

   return 0;
}

File: Olympiad/COCI/official/2008/contest4_solutions/poklon.pas
program poklon;

const MAX = 1000000;
const MAXN = 100000;

type interval = record lo, hi : longint; end;

function cmp( A, B : interval ) : boolean;
begin
   cmp := (A.lo > B.lo) or (A.lo = B.lo) and (A.hi < B.hi);
end;

var
   n : longint;
   intervals, tmp : array[1..MAXN] of interval;
   dp, how : array[1..MAXN] of longint;
   fenwick : array[0..MAX] of longint;
   i, x, ret : longint;

function query( x : longint ) : longint;
var
   i, ret : longint;
begin
   ret := 0;

   i := x;
   while i > 0 do begin
      if (fenwick[i] <> 0) and ((ret = 0) or (dp[fenwick[i]] > dp[ret])) then
         ret := fenwick[i];
      i := i - (i and -i);
   end;

   query := ret;
end;

procedure update( x, v : longint );
var
   i : longint;
begin
   i := x;
   while i <= MAX do begin
      if (fenwick[i] = 0) or (dp[v] > dp[fenwick[i]]) then
         fenwick[i] := v;
      i := i + (i and -i);
   end;
end;

procedure mergesort( lo, hi : longint );
var
   mid : longint;
   i, j, k : longint;
begin
   if lo <> hi then begin
      mid := (lo+hi) div 2;
      mergesort( lo, mid );
      mergesort( mid+1, hi );

      i := lo;
      j := mid+1;
      k := lo;
      while (i <= mid) and (j <= hi) do begin
         if cmp( intervals[i], intervals[j] ) then begin
            tmp[k] := intervals[i];
            i := i + 1;
         end else begin
            tmp[k] := intervals[j];
            j := j + 1;
         end;
         k := k + 1;
      end;
      while i <= mid do begin
         tmp[k] := intervals[i];
         i := i + 1;
         k := k + 1;
      end;
      while j <= hi do begin
         tmp[k] := intervals[j];
         j := j + 1;
         k := k + 1;
      end;
      for i := lo to hi do intervals[i] := tmp[i];
   end;
end;

begin
   readln( n );

   for i := 1 to n do readln( intervals[i].lo, intervals[i].hi );

   mergesort( 1, n );

   ret := 1;
   for i := 1 to n do begin;
      dp[i] := 1;
      how[i] := 0;

      x := query( intervals[i].hi );

      if x <> 0 then begin
         dp[i] := dp[x] + 1;
         how[i] := x;
      end;

      update( intervals[i].hi, i );

      if dp[i] > dp[ret] then ret := i;
   end;

   writeln( dp[ret] );
   while ret <> 0 do begin
      writeln( intervals[ret].lo, ' ', intervals[ret].hi );
      ret := how[ret];
   end;
end.
